design/service-apis-implementation.md

```diff
@@ -0,0 +1,326 @@
+
+#### Conflict Resolution and Route merging
+
+HTTPRoutes may be merged into a single Envoy config by Contour as long as:
+- the Hostnames match exactly and the TLS Config matches
+- the HTTPRouteRules are different
+
+In the event that two HTTPRoutes match on Hostname and RouteTLSConfig, and have matching HTTPRouteRules (`rules` stanza),
+then the oldest HTTPRoute's HTTPRouteRules should win.
+The not accepted HTTPRoute must have its `Admitted` Condition changed to `status: false` with a reason field to explain why, possibly including the name of the other resource.
+
+If Hostnames match exactly, but TLS Config does not, that is a conflict and the oldest-object-wins rule applies also.
+That is, you can't specify a different certificate for the same hostname.
+
+### TLSRoute
+
+TLSRoute implementation is quite straightforward, only allows simple routing based on SNI.
+Contour will implement this object as per its API spec.
+### Other concerns
+
+## Alternatives Considered
+
+### Contour watches multiple Gateways
+The only main alternative considered for this design was making Contour possibly responsible for more than one Gateway.
+In this model, Contour would be configured with a `controller` string, and watch all GatewayClasses for that string in the `spec.controller` field.
+Contour would also watch all Gateways, and for Gateways with a matching `spec.gatewayClassName`, would use those as a basis for looking for routes.
+
+The reason this model wasn't chosen was that it's difficult to define how you could merge multiple Gateway objects into a single Envoy installation.
+Model-wise, it seems to be intended that a Gateway represents the actual thing that takes traffic and transforms it so it can get to the requested Pods.
+In Contour's case, this is Envoy, and it's extremely difficult to design a model where a single Envoy deployment could handle multiple different Gateway specs.
+
+Also, this model makes interacting with the Contour Operator much harder.
+In the model we've chosen, the Operator takes the Gateway as a request for an Envoy installation, with matching Contour installation.
+As part of that, Contour is requested to look at one and only one Gateway, which contradicts the "watches multiple Gateways" model.
+
+## Implementation
+
+There is currently experimental support for *reading* the Service-APIs objects in Contour, we will add full support in something like this order:
+- Add support for configuring Gateway to the config file
+- Deprecate `--expermental-service-apis` command-line flag, so that it does nothing but log a referral to the new config item.
+- Implement the Service-APIs processor as a single DAG processor.
+- Release a version with Service-APIs support still marked experimental.
+- Remove the experimental tag once we're happy with the support.
```

Comment:
Might be worth adding some notes about how before this stage we will try to evaluate our implementation and work with upstream to work out issues, conformance testing, etc.